lib/shared/sdk/usbboot/index.js

```diff
@@ -0,0 +1,318 @@
+    return protocol.write(device, endpoint, bootCodeBuffer)
+  }).then(() => {
+    debug('Reading return code from device')
+    return protocol.read(device, protocol.RETURN_CODE_LENGTH)
+  }).then((data) => {
+    const returnCode = data.readInt32LE()
+    debug(`Received return code: ${returnCode}`)
+
+    if (returnCode !== protocol.RETURN_CODE_SUCCESS) {
+      throw new Error(`Couldn't write the bootcode, got return code ${returnCode} from device`)
+    }
+  })
+}
+
+const startFileServer = (device, endpoint, files) => {
+  debug('Listening for file messages')
+  return protocol
+    .read(device, protocol.FILE_MESSAGE_LENGTH)
+    .then(protocol.parseFileMessageBuffer)
+
+    // We get this error message when reading a command
+    // from the device when the communication has ended
+    .catch({
+      message: 'LIBUSB_TRANSFER_STALL'
+    }, (error) => {
+      debug(`Got ${error.message} when reading a command, assuming everything is done`)
+      return {
+        command: protocol.FILE_MESSAGE_COMMAND_DONE
+      }
+    })
+
+    .then((fileMessage) => {
+      debug(`Received message: ${fileMessage.command} -> ${fileMessage.fileName}`)
+
+      if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_DONE) {
+        debug('Done')
+        return Bluebird.resolve()
+      }
+
+      return Bluebird.try(() => {
+        if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_GET_FILE_SIZE) {
+          debug(`Getting the size of ${fileMessage.fileName}`)
+          const fileSize = _.get(files, [ fileMessage.fileName, 'length' ])
+
+          if (_.isNil(fileSize)) {
+            debug(`Couldn't find ${fileMessage.fileName}`)
+            debug('Sending error signal')
+            return protocol.sendErrorSignal(device)
+          }
+
+          debug(`Sending size: ${fileSize}`)
+          return protocol.sendBufferSize(device, fileSize)
+        }
+
+        if (fileMessage.command === protocol.FILE_MESSAGE_COMMAND_READ_FILE) {
+          debug(`Reading ${fileMessage.fileName}`)
+          const fileBuffer = _.get(files, [ fileMessage.fileName ])
+
+          if (_.isNil(fileBuffer)) {
+            debug(`Couldn't find ${fileMessage.fileName}`)
+            debug('Sending error signal')
+            return protocol.sendErrorSignal(device)
+          }
+
+          return protocol.write(device, endpoint, fileBuffer)
+        }
+
+        return Bluebird.reject(new Error(`Unrecognized command: ${fileMessage.command}`))
+      }).then(() => {
+        debug('Starting again')
+        return startFileServer(device, endpoint, files)
+      })
+    })
+}
+
+exports.flash = (device, options = {}) => {
+  // We need to open the device in order to access _configDescriptor
+  debug(`Opening device: ${device.name}`)
+  device.open()
+
+  // Handle 2837 where it can start with two interfaces, the first
+  // is mass storage the second is the vendor interface for programming
+  const addresses = {}
```

Comment:
Does `address` make it better? I honestly can't think of a better name, but I'm open to suggestions :)